Translation.md

```diff
@@ -371,6 +371,159 @@ https://github.com/saltiniroberto/ssf/blob/7ea6e18093d9da3154b4e396dd435549f687e
 We translate this by mapping the function over its `DOMAIN`.
 
 
+## Meta-rules
+
+In order to facilitate translation to a fragment supported by Apalache, we need to introduce a set of TLA-to-TLA rules, which allow us to formulate translations from Python to TLA+ first in the intuitive way, potentially introducing constructs like recursion, which we then pair with a TLA-to-TLA rule, ending in a supported fragment. 
+
+### Bounded primitive recursion rule
+
+Assume we are given a `RECURSIVE` operator `Op` of arity `1`, s.t.
+```
+RECURSIVE Op(_)
+\* @type (a) => a;
+Op(x) ==
+  IF P(x)
+  THEN e
+  ELSE G(x, Op(b(x))
+```
+
+If, for every `x`, there exists a sequence `x = v_1, ..., v_n`, such that
+  - `n + 1` is bounded by some constant `N` (for any particular choice of `x` the constant may be different)
+  - `P(v_n)` holds
+  - `v_{i+1} = b(v_i)`
+
+we can define:
+```
+\* @type (a, Int) => Seq(a);
+Chain(x, N) ==
+  LET 
+    \* @type: (Seq(a), Int) => Seq(a);
+    step(seq, i) == 
+      IF i > Len(seq) \/ P(seq[1])
+      THEN seq
+      ELSE <<b(seq[1])>> \o seq \* Alternatively, we can append here and reverse the list at the end
+  IN ApaFoldSeqLeft( step, <<x>>, MkSeq(N, LAMBDA i: i) )
+```
+We know the above assumptions hold, if `P(Chain(x)[1])` holds, as it returns the sequence `<<v_n, ..., x>>` under those circumstances.
+
+Using `Chain` we can define a fold-based non-recursive operator `Op^`, such that `Op^(x) = Op(x)` under the above assumptions:
+
+```
+\* @type (a, Int) => a;
+NonrecursiveOp(x, N) ==
+LET chain == Chain(x, N) IN
+LET step(cumul, v) == G(v, cumul) IN
+ApaFoldSeqLeft( step, e, Tail(chain) )
+```
+
+Then, `Op^(x) = NonrecursiveOp(x, N)`. Alternatively
+
+```
+\* @type (a, Int) => a;
+NonrecursiveOp(x, N) ==
+LET chain == Chain(x, N) IN
+LET step(cumul, v) == G(v, cumul) IN
+IF ~P(chain[1])
+THEN CHOOSE x \in {}: TRUE
+ELSE ApaFoldSeqLeft( step, e, Tail(chain) )
+```
+
+In the special case where `G` is associative, that is, `G(a, G(b, c)) = G(G(a, b), c)` for all `a,b,c`, we can make the entire translation more optimized, and single-pass. Since `NonrecursiveOp(x,N)`, for sufficiently large `N`, computes 
+```
+G(v_1, G(v_2, ... (G(v_{n-2}, G(v_{n-1}, e)))))
+```
+and `G` is associative by assumption, then computing
+```
+G(G(G(G(v_1, v_2), ...), v_{n-1}), e)
+```
+should give us the same value. However, this computation can be done in a single pass:
+```
+NonrecursiveOpForAssociative(x, N) ==
+  IF P(x)
+  THEN e
```

Comment:
Unfortunately no, since this is the only way you get the correct result when `P(x)` already holds.
You would have to start the fold with `<<e, x>>` in that case, since 
```
          IF P(lastElemInSeq)
          THEN pair
```
would give you the initial value forever. However, you can't start the fold with `<<e,x>>`, because in the general case, the "step 1" value you need to compute is `G(v_1, v_2)`, and if you started with `<<e,x>>` you would get `G(e, v_2)`, which is incorrect.

Reply:
Oh right, I missed that.
Thanks for clarifying!